sonar-orchestrator/src/main/java/com/sonar/orchestrator/build/ScannerForMSBuildInstaller.java

```diff
@@ -58,54 +60,61 @@ public ScannerForMSBuildInstaller(Locators locators) {
   }
 
   /**
-   * Installs an ephemeral Scanner for MSBuild and returns the path to the exe to execute
+   * Installs an ephemeral Scanner for .NET and returns the path to the exe to execute.
    */
   public File install(@Nullable Version scannerVersion, @Nullable Location location, File toDir, boolean useDotNetCore) {
-    if (location != null) {
-      return installFromLocation(location, toDir, useDotNetCore);
-    } else if (scannerVersion != null) {
-      return install(scannerVersion, toDir, useDotNetCore);
+    if (location == null) {
+      scannerVersion = scannerVersion == null
+              ? Version.create(DEFAULT_SCANNER_VERSION)
+              : scannerVersion;
+      PackageDetails packageDetails = packageDetailsFactory.create(scannerVersion, useDotNetCore);
+      return install(scannerVersion, toDir, packageDetails);
     } else {
-      return install(Version.create(DEFAULT_SCANNER_VERSION), toDir, useDotNetCore);
+      return installFromLocation(location, toDir, useDotNetCore);
     }
   }
 
   private File installFromLocation(Location location, File toDir, boolean useDotNetCore) {
-    clearCachedSnapshot(null, toDir, useDotNetCore);
-    if (!isInstalled(null, toDir, useDotNetCore)) {
-      LOG.info("Installing Scanner for MSBuild from {}", location);
-      File zipFile = locators.locate(location);
-
-      if (zipFile == null || !zipFile.exists()) {
-        throw new IllegalArgumentException("File doesn't exist: " + zipFile);
-      }
-      doInstall(zipFile, toDir, null, useDotNetCore);
+    // When installing from specific location, the folder is always the same and the cache directory is always deleted.
+    String scannerFolderName = "sonar-scanner";
+    String executableFileName = useDotNetCore ? "SonarScanner.MSBuild.dll" : "SonarScanner.MSBuild.exe";
+    clearCachedSnapshot(toDir, scannerFolderName);
+
+    LOG.info("Installing Scanner for .NET from {}", location);
+    File zipFile = locators.locate(location);
+    if (zipFile == null || !zipFile.exists()) {
+      throw new IllegalArgumentException("File doesn't exist: " + zipFile);
     }
-    return locateInstalledScript(null, toDir, useDotNetCore);
+    doInstall(zipFile, toDir, null, scannerFolderName);
+    return locateInstalledScript(toDir, scannerFolderName, executableFileName);
   }
 
-  private File install(Version scannerVersion, File toDir, boolean useDotNetCore) {
-    clearCachedSnapshot(scannerVersion, toDir, useDotNetCore);
-    if (!isInstalled(scannerVersion, toDir, useDotNetCore)) {
+  private File install(Version scannerVersion, File toDir, PackageDetails packageDetails) {
+    if (scannerVersion.isSnapshot()) {
```

Comment:
Looking at what `Version.isSnapshot()` does, I'm not sure it will ever be true, so I won't bother with this check. But you know better the versionning of this scanner, I let you decide.

Reply:
Indeed, looking at repox, we never have snapshots generated. Each PR has it's own version. I can remove the if block and the `clearCachedSnapshot` call.